VideoLocker/src/main/java/org/edx/mobile/view/CourseOutlineFragment.java

```diff
@@ -183,4 +193,58 @@ public void reloadList(){
             adapter.reloadData();
         }
     }
+
+    private void updateRowSelection(String lastAccessedId){
+        if (!TextUtils.isEmpty(lastAccessedId)) {
+            final int selectedItemPosition = adapter.getPositionByItemId(lastAccessedId);
+            if (selectedItemPosition != -1) {
+                listView.setItemChecked(selectedItemPosition, true);
+                listView.postDelayed(new Runnable() {
+                    @Override
+                    public void run() {
+                        listView.smoothScrollToPosition(selectedItemPosition);
+                    }
+                }, AUTOSCROLL_DELAY_MS);
+            }
+        }
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+        switch (requestCode) {
+            // If user has navigated to a different unit, then we need to rearrange
+            // the activity stack to point to it.
+            case REQUEST_SHOW_COURSE_UNIT_DETAIL: {
+                switch (resultCode) {
+                    case Activity.RESULT_OK: {
+                        CourseComponent outlineComp = courseManager.getComponentById(
+                                courseData.getCourse().getId(), courseComponentId);
+                        String leafCompId = (String) data.getSerializableExtra(Router.EXTRA_COURSE_COMPONENT_ID);
+                        CourseComponent leafComp = courseManager.getComponentById(
+                                courseData.getCourse().getId(), leafCompId);
+                        BlockPath outlinePath = outlineComp.getPath();
+                        BlockPath leafPath = leafComp.getPath();
+                        int outlinePathSize = outlinePath.getPath().size();
+                        if (!outlineComp.equals(leafPath.get(outlinePathSize - 1))) {
+                            getActivity().setResult(Activity.RESULT_OK, data);
+                            getActivity().finish();
+                        } else {
+                            for (int i = outlinePathSize + 1; ; i += 2) {
+                                CourseComponent nextComp = leafPath.get(i);
+                                if (nextComp == null || !nextComp.isContainer()) {
+                                    updateRowSelection(leafCompId);
```

Comment:
`updateRowSelection()` should only be called from the final instance of the `CourseOutlineFragment` started here. If there is nothing to be started, then it can be called directly from here. However, calling it before breaking from the loop ensures that it will be called from the base instance as well.

Here's an adapted version of the code in the `else` section that should have the proper logic:
```` java
int leafPathSize = leafPath.getPath().size();
if (outlinePathSize == leafPathSize - 1) {
    updateRowSelection(leafCompId);
} else {
    for (int i = outlinePathSize + 3;; i += 2) {
        CourseComponent nextComp = leafPath.get(i);
        if (nextComp == null || !nextComp.isContainer()) {
            updateRowSelection(leafCompId);
            break;
        }
        environment.getRouter().showCourseContainerOutline(CourseOutlineFragment.this,
                REQUEST_SHOW_COURSE_UNIT_DETAIL, courseData, nextComp.getId(), leafCompId);
    }
    getActivity().overridePendingTransition(R.anim.slide_in_from_start, R.anim.slide_out_to_end);
}
````